docs/architecture/adr-058-auto-generated-cli.md

```diff
@@ -0,0 +1,83 @@
+go module - for automatically generating CLI commands based on protobuf definitions to either replace or complement
+handwritten CLI commands. This will mean that when developing a module, it will be possibly to skip both writing and
+testing CLI commands as that can all be taken care of by the framework.
+
+The basic design for automatically generating CLI commands is to:
+* create one CLI command for each `rpc` method in a protobuf `Query` or `Msg` service
+* create a CLI flag for each field in the `rpc` request type
+* for `query` commands call gRPC and print the response as protobuf JSON or YAML
+* for `tx` commands, create a transaction and apply common transaction flags
+
+In order to make the auto-generated CLI as easy to use (or easier) than handwritten CLI, we need to do custom handling
+of specific protobuf field types so that the input format is easy for humans:
+* `Coin`, `Coins`, `DecCoin`, and `DecCoins` should be input using the existing format (i.e. `1000uatom`)
+* it should be possible to specify an address using either the bech32 address string or a named key in the keyring
+* `Timestamp` and `Duration` should accept strings like `2001-01-01T00:00:00Z` and `1h3m` respectively
+* pagination should be handled with flags like `--page-limit`, `--page-offset`, etc.
+* it should be possible to customize any other protobuf type either via its message name or a `cosmos_proto.scalar` annotation
+
+At a basic level it should be possible to generate a command for a single `rpc` method as well as all the commands for
+a whole protobuf `service` definition. It should be possible to mix and match auto-generated and handwritten commands.
+
+## Consequences
+
+### Backwards Compatibility
+
+Existing modules can mix and match auto-generated and handwritten CLI commands so it is up to them as to whether they
+make breaking changes by replacing handwritten commands with slightly different auto-generated ones.
+
+### Positive
+
+* module developers will not need to write CLI commands
+* module developers will not need to test CLI commands
+
+### Negative
+
+### Neutral
+
+## Further Discussions
+
+At some point in the future, we should allow customization of the short and long usage strings for commands as
+well as allowing shorthand aliases for flags (ex. `-a` for `--amount`). We may also consider a way to specify that
+certain arguments should be positional parameters rather than flags if the UX merits this. It is an [open discussion](https://github.com/cosmos/cosmos-sdk/pull/11725#issuecomment-1108676129)
+as to whether these customizations options should live the .proto files themselves, in separate config files, or
+in directly in code. Providing the options in .proto files would allow a dynamic client to automatically generate
+CLI commands on the fly. However, that may pollute the .proto files themselves with information that is only relevant
+for a small subset of users.
+
```

Comment:
Should we add a note that lens, which aims to be a chain-agnostic CLI client, could benefit largely from this?